app/src/androidTest/java/br/com/catbag/gifreduxsample/ui/giflist/GifListActivityTest.java

```diff
@@ -212,105 +215,106 @@ public void whenGifDownloadFailTest() {
         int gifListSize = 5;
-        dispatchFakeReduceAction(createStateFromGifs(createFakeGifList(gifListSize)));
+        helper.dispatchFakeAppState(buildAppState(createFakeGifList(gifListSize)));
 
         mActivityTestRule.launchActivity(new Intent());
 
-        FeedComponent feed = (FeedComponent) getActivity().findViewById(R.id.feed);
-        RecyclerView recyclerView = (RecyclerView) feed.getChildAt(0);
-        recyclerView.setId(View.generateViewId());
-
-        onView(withId(recyclerView.getId())).perform(scrollToPosition(gifListSize-1));
-
+        //Initiate lock on first adapter GifComponent
         UiTestLocker locker = new UiTestLocker(getGifComponent(0));
 
-        onView(withId(recyclerView.getId()))
-                .perform(RecyclerViewActions.actionOnItemAtPosition(0, click()));
+        //Scroll to bottom
+        onView(withId(getRecyclerView().getId())).perform(scrollToPosition(gifListSize-1));
+
+        //Scroll to top and click on first adapter GifComponent
+        onView(withId(getRecyclerView().getId()))
+                .perform(actionOnItemAtPosition(0, click()));
 
         locker.registerIdlingResource();
 
-        int firstViewId = View.generateViewId();
-        FrameLayout frame = ((FrameLayout)recyclerView.getChildAt(0));
-        View firstView = frame.findViewById(R.id.gif_image);
-        firstView.setId(firstViewId);
+        //Get the first GifImageView and set a unique id
+        int tempId = generateViewId();
+        getGifImgViewAt(R.id.gif_image, 0).setId(tempId);
 
-        onView(withId(firstViewId))
-                .check(matches(withPlayingGifDrawable()));
+        //Checks if it has a playing status
+        onView(withId(tempId)).check(matches(withPlayingGifDrawable()));
 
-        firstView.setId(R.id.gif_image);
+        //clean test modifications
+        getGifImgViewAt(tempId, 0).setId(R.id.gif_image);
         locker.unregisterIdlingResource();
     }
 
-    private void dispatchFakeReduceAction(AppState state) {
-        getFluxxan().getDispatcher()
-                .dispatch(new Action(FakeReducer.FAKE_REDUCE_ACTION, state));
+    private RecyclerView getRecyclerView() {
+        FeedComponent feed = (FeedComponent) getActivity().findViewById(R.id.feed);
+        RecyclerView recyclerView = (RecyclerView) feed.getChildAt(0);
+        recyclerView.setId(generateViewId());
+        return recyclerView;
     }
 
-    private GifListActivity getActivity(){
-        return mActivityTestRule.getActivity();
+    private GifComponent getGifComponent(int screenPos) {
+        FrameLayout frameLayout = (FrameLayout) getRecyclerView().getChildAt(screenPos);
+        return (GifComponent) frameLayout.getChildAt(0);
+    }
+
+    private GifImageView getGifImgViewAt(int viewId, int screenPos) {
+        FrameLayout frameLayout = (FrameLayout) getRecyclerView().getChildAt(screenPos);
+        return (GifImageView) frameLayout.findViewById(viewId);
```

Comment:
You can make use of `getGifComponent(screenPos).findViewById(viewId)` to avoid duplicate frame layout search.